flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/ReadWriteTableITCase.java

```diff
@@ -19,73 +19,242 @@
+            }
+        }
+    }
 
     public ReadWriteTableITCase(
             RuntimeExecutionMode executionMode,
             String tableName,
             boolean enableLogStore,
-            boolean hasPk,
-            @Nullable Boolean duplicate,
+            String helperTableDdl,
+            String managedTableDdl,
+            String insertQuery,
+            String selectQuery,
             ExpectedResult expectedResult) {
         super(executionMode, tableName, enableLogStore, expectedResult);
-        this.hasPk = hasPk;
-        this.duplicate = duplicate;
+        this.helperTableDdl = helperTableDdl;
+        this.managedTableDdl = managedTableDdl;
+        this.insertQuery = insertQuery;
+        this.selectQuery = selectQuery;
+    }
+
+    @Parameterized.Parameters(
+            name =
+                    "executionMode-{0}, tableName-{1}, "
+                            + "enableLogStore-{2}, helperTableDdl-{3}, managedTableDdl-{4}, "
+                            + "insertQuery-{5}, selectQuery-{6}, expectedResult-{7}")
+    public static List<Object[]> data() {
+        List<Object[]> specs = prepareReadWriteTestSpecs();
+        specs.addAll(prepareBatchWriteStreamingReadTestSpecs());
+        specs.addAll(prepareStreamingWriteBatchReadTestSpecs());
+        specs.addAll(prepareOverwriteTestSpecs());
+        return specs;
     }
 
     @Override
-    public void after() {
-        tEnv.executeSql("DROP TABLE `source_table`");
-        super.after();
+    protected void prepareEnv() {
+        tEnv.executeSql(helperTableDdl);
+        tEnv.executeSql(managedTableDdl);
     }
 
     @Test
-    public void testReadWriteNonPartitioned() throws Exception {
-        String statement =
-                String.format("INSERT INTO %s \nSELECT * FROM `source_table`", tableIdentifier);
+    public void testSequentialWriteRead() throws Exception {
+        logTestSpec(
+                executionMode,
+                tableIdentifier,
+                enableLogStore,
+                helperTableDdl,
+                managedTableDdl,
+                insertQuery,
+                selectQuery,
+                expectedResult);
         if (expectedResult.success) {
-            tEnv.executeSql(statement).await();
-            TableResult result =
-                    tEnv.executeSql(String.format("SELECT * FROM %s", tableIdentifier));
+            tEnv.executeSql(insertQuery).await();
+            TableResult result = tEnv.executeSql(selectQuery);
             List<Row> actual = new ArrayList<>();
+
             try (CloseableIterator<Row> iterator = result.collect()) {
-                while (iterator.hasNext()) {
-                    actual.add(iterator.next());
+                if (env.getConfiguration().get(ExecutionOptions.RUNTIME_MODE)
+                        == RuntimeExecutionMode.STREAMING) {
+                    ExecutorService executorService = Executors.newSingleThreadExecutor();
```

Comment:
Why use a `ExecutorService`?

Reply:
> Why use a `ExecutorService`?

Because in streaming mode, read may get blocked if the expected records num not equal with actual

Reply:
Good idea, we do not want hang forever util github testing timeout.
I create a PR to introduce the timeout helper: https://github.com/apache/flink-table-store/pull/60